cpp: libVGELib.so build error

```
$ time make
[ 85%] Building CXX object CMakeFiles/VGELib.dir/vgelib/pipeline.cpp.o
/home/u/Desktop/VGE/cpp/vgelib/pipeline.cpp: In member function ‘void vge::GraphicsPipeline::Create(vge::RenderPass*)’:
/home/u/Desktop/VGE/cpp/vgelib/pipeline.cpp:74:102: error: ambiguous overload for ‘operator=’ (operand types are ‘vk::Pipeline’ and ‘vk::ResultValue<vk::Pipeline>’)
   74 |  _pipeline = _dev->get_device().createGraphicsPipeline(nullptr, gpci, allocator, _dev->get_dispatch());
      |                                                                                                      ^
In file included from /home/u/Desktop/VGE/cpp/vgelib/vgelib.hpp:11,
                 from /home/u/Desktop/VGE/cpp/vgelib/pipeline.cpp:1:
/usr/include/vulkan/vulkan.hpp:26417:9: note: candidate: ‘constexpr vk::Pipeline& vk::Pipeline::operator=(const vk::Pipeline&)’
26417 |   class Pipeline
      |         ^~~~~~~~
/usr/include/vulkan/vulkan.hpp:26417:9: note: candidate: ‘constexpr vk::Pipeline& vk::Pipeline::operator=(vk::Pipeline&&)’
/home/u/Desktop/VGE/cpp/vgelib/pipeline.cpp: In member function ‘void vge::ComputePipeline::Create()’:
/home/u/Desktop/VGE/cpp/vgelib/pipeline.cpp:175:101: error: ambiguous overload for ‘operator=’ (operand types are ‘vk::Pipeline’ and ‘vk::ResultValue<vk::Pipeline>’)
  175 |  _pipeline = _dev->get_device().createComputePipeline(nullptr, cpci, allocator, _dev->get_dispatch());
      |                                                                                                     ^
In file included from /home/u/Desktop/VGE/cpp/vgelib/vgelib.hpp:11,
                 from /home/u/Desktop/VGE/cpp/vgelib/pipeline.cpp:1:
/usr/include/vulkan/vulkan.hpp:26417:9: note: candidate: ‘constexpr vk::Pipeline& vk::Pipeline::operator=(const vk::Pipeline&)’
26417 |   class Pipeline
      |         ^~~~~~~~
/usr/include/vulkan/vulkan.hpp:26417:9: note: candidate: ‘constexpr vk::Pipeline& vk::Pipeline::operator=(vk::Pipeline&&)’
make[2]: *** [CMakeFiles/VGELib.dir/build.make:147: CMakeFiles/VGELib.dir/vgelib/pipeline.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:134: CMakeFiles/VGELib.dir/all] Error 2
make: *** [Makefile:149: all] Error 2
```

The error was encountered at rev 15f418d41aa2c4d90039c4b7c51ce6c2231eecdf, using GCC 10.2.0 on Arch Linux, and Vulkan header version 1.2.172. Let me know if you need any additional details to troubleshoot.

```
$ uname -a
Linux x1 5.11.7-arch1-1 #1 SMP PREEMPT Wed, 17 Mar 2021 16:59:58 +0000 x86_64 GNU/Linux
```

```
$ gcc -v
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-pc-linux-gnu/10.2.0/lto-wrapper
Target: x86_64-pc-linux-gnu
Configured with: /build/gcc/src/gcc/configure --prefix=/usr --libdir=/usr/lib --libexecdir=/usr/lib --mandir=/usr/share/man --infodir=/usr/share/info --with-bugurl=https://bugs.archlinux.org/ --enable-languages=c,c++,ada,fortran,go,lto,objc,obj-c++,d --with-isl --with-linker-hash-style=gnu --with-system-zlib --enable-__cxa_atexit --enable-cet=auto --enable-checking=release --enable-clocale=gnu --enable-default-pie --enable-default-ssp --enable-gnu-indirect-function --enable-gnu-unique-object --enable-install-libiberty --enable-linker-build-id --enable-lto --enable-multilib --enable-plugin --enable-shared --enable-threads=posix --disable-libssp --disable-libstdcxx-pch --disable-libunwind-exceptions --disable-werror gdc_include_dir=/usr/include/dlang/gdc
Thread model: posix
Supported LTO compression algorithms: zlib zstd
gcc version 10.2.0 (GCC) 
```

```
$ pacman -Qi vulkan-headers
Name            : vulkan-headers
Version         : 1:1.2.172-1
```

While here, I might as well post a warning that was encountered when compiling `vgelib/command.cpp`. Of course, this is less critical to fix, but fixing warnings for the default build makes it easier to spot real errors.

```
[ 70%] Building CXX object CMakeFiles/VGELib.dir/vgelib/command.cpp.o
/home/u/Desktop/VGE/cpp/vgelib/command.cpp: In member function ‘void vge::Queue::submit(vge::Command*, vge::SubmitInfo**, size_t, vk::PipelineStageFlags, vge::SubmitInfo*&)’:
/home/u/Desktop/VGE/cpp/vgelib/command.cpp:52:32: warning: ignoring return value of ‘vk::Result vk::Device::resetFences(uint32_t, const vk::Fence*, const Dispatch&) const [with Dispatch = vk::DispatchLoaderDynamic; uint32_t = unsigned int]’, declared with attribute ‘nodiscard’ [-Wunused-result]
   52 |  _dev->get_device().resetFences(1, &(cmd->_fence), _dev->get_dispatch());
      |  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /home/u/Desktop/VGE/cpp/vgelib/vgelib.hpp:11,
                 from /home/u/Desktop/VGE/cpp/vgelib/command.cpp:1:
/usr/include/vulkan/vulkan.hpp:88681:49: note: declared here
88681 |   VULKAN_HPP_NODISCARD VULKAN_HPP_INLINE Result Device::resetFences( uint32_t fenceCount, const VULKAN_HPP_NAMESPACE::Fence* pFences, Dispatch const & d  ) const VULKAN_HPP_NOEXCEPT
      |                                                 ^~~~~~
/home/u/Desktop/VGE/cpp/vgelib/command.cpp:53:15: warning: ignoring return value of ‘vk::Result vk::Queue::submit(uint32_t, const vk::SubmitInfo*, vk::Fence, const Dispatch&) const [with Dispatch = vk::DispatchLoaderDynamic; uint32_t = unsigned int]’, declared with attribute ‘nodiscard’ [-Wunused-result]
   53 |  _queue.submit(1, &si, cmd->_fence, _dev->get_dispatch());
      |  ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /home/u/Desktop/VGE/cpp/vgelib/vgelib.hpp:11,
                 from /home/u/Desktop/VGE/cpp/vgelib/command.cpp:1:
/usr/include/vulkan/vulkan.hpp:92248:49: note: declared here
92248 |   VULKAN_HPP_NODISCARD VULKAN_HPP_INLINE Result Queue::submit( uint32_t submitCount, const VULKAN_HPP_NAMESPACE::SubmitInfo* pSubmits, VULKAN_HPP_NAMESPACE::Fence fence, Dispatch const & d  ) const VULKAN_HPP_NOEXCEPT
      |                                                 ^~~~~
/home/u/Desktop/VGE/cpp/vgelib/command.cpp: In member function ‘virtual void vge::Command::Dispose()’:
/home/u/Desktop/VGE/cpp/vgelib/command.cpp:63:35: warning: ignoring return value of ‘vk::Result vk::Device::waitForFences(uint32_t, const vk::Fence*, vk::Bool32, uint64_t, const Dispatch&) const [with Dispatch = vk::DispatchLoaderDynamic; uint32_t = unsigned int; vk::Bool32 = unsigned int; uint64_t = long unsigned int]’, declared with attribute ‘nodiscard’ [-Wunused-result]
   63 |   _dev->get_device().waitForFences(1, &_fence, 1, MaxTimeout, _dev->get_dispatch());
      |   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /home/u/Desktop/VGE/cpp/vgelib/vgelib.hpp:11,
                 from /home/u/Desktop/VGE/cpp/vgelib/command.cpp:1:
/usr/include/vulkan/vulkan.hpp:88893:49: note: declared here
88893 |   VULKAN_HPP_NODISCARD VULKAN_HPP_INLINE Result Device::waitForFences( uint32_t fenceCount, const VULKAN_HPP_NAMESPACE::Fence* pFences, VULKAN_HPP_NAMESPACE::Bool32 waitAll, uint64_t timeout, Dispatch const & d  ) const VULKAN_HPP_NOEXCEPT
      |                                                 ^~~~~~
/home/u/Desktop/VGE/cpp/vgelib/command.cpp: In member function ‘void vge::Command::Wait()’:
/home/u/Desktop/VGE/cpp/vgelib/command.cpp:216:34: warning: ignoring return value of ‘vk::Result vk::Device::waitForFences(uint32_t, const vk::Fence*, vk::Bool32, uint64_t, const Dispatch&) const [with Dispatch = vk::DispatchLoaderDynamic; uint32_t = unsigned int; vk::Bool32 = unsigned int; uint64_t = long unsigned int]’, declared with attribute ‘nodiscard’ [-Wunused-result]
  216 |  _dev->get_device().waitForFences(1, &_fence, true, MaxTimeout, _dev->get_dispatch());
      |  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /home/u/Desktop/VGE/cpp/vgelib/vgelib.hpp:11,
                 from /home/u/Desktop/VGE/cpp/vgelib/command.cpp:1:
/usr/include/vulkan/vulkan.hpp:88893:49: note: declared here
88893 |   VULKAN_HPP_NODISCARD VULKAN_HPP_INLINE Result Device::waitForFences( uint32_t fenceCount, const VULKAN_HPP_NAMESPACE::Fence* pFences, VULKAN_HPP_NAMESPACE::Bool32 waitAll, uint64_t timeout, Dispatch const & d  ) const VULKAN_HPP_NOEXCEPT
      |                                                 ^~~~~~
```

--- a/cpp/CMakeLists.txt
+++ b/cpp/CMakeLists.txt
@@ -12,7 +12,7 @@ if (CMAKE_CXX_COMPILER_ID MATCHES "MSVC")
 else()
   set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++17")
 endif()
-set(VULKAN_SDK "D:/Libs/VulkanSDK/1_1_101" CACHE PATH "Vulkan SDK path")
+set(VULKAN_SDK "" CACHE PATH "Vulkan SDK path")
 
 
 
@@ -21,7 +21,11 @@ set(GLFW_BUILD_TESTS OFF CACHE BOOL "" FORCE)
 set(GLFW_BUILD_EXAMPLES OFF CACHE BOOL "" FORCE)
 add_subdirectory("third_party/glfw")
 
-include_directories(${PROJECT_SOURCE_DIR} "third_party/glfw/include" "${VULKAN_SDK}/include" )
+include_directories(${PROJECT_SOURCE_DIR} "third_party/glfw/include")
+
+if(NOT VULKAN_SDK STREQUAL "")
+    include_directories("${VULKAN_SDK}/include")
+endif()
 
 if(NOT CMAKE_DEBUG_POSTFIX)
   set(CMAKE_DEBUG_POSTFIX d)
@@ -34,7 +38,11 @@ if(CMAKE_SYSTEM_NAME MATCHES "Windows")
     set(VULKAN_LIB "${VULKAN_SDK}/lib/vulkan-1.lib")
     set(VGELIB_OS_SRC "vgelib/common/stb_image.cpp")
 else()
-    set(VULKAN_LIB "${VULKAN_SDK}/lib/libvulkan.so")
+    if(NOT VULKAN_SDK STREQUAL "")
+        set(VULKAN_LIB "${VULKAN_SDK}/lib/libvulkan.so")
+    else()
+        set(VULKAN_LIB "/usr/lib/libvulkan.so")
+    endif()
     set(VGELIB_OS_SRC "vgelib/common/stb_image.cpp" )
 endif()
 

--- a/cpp/vgelib/pipeline.cpp
+++ b/cpp/vgelib/pipeline.cpp
@@ -71,7 +71,7 @@ void vge::GraphicsPipeline::Create(RenderPass* renderPass)
 	pmsci.minSampleShading = 1;
 	pmsci.rasterizationSamples = vk::SampleCountFlagBits::e1;
 	gpci.pMultisampleState = &pmsci;
-	_pipeline = _dev->get_device().createGraphicsPipeline(nullptr, gpci, allocator, _dev->get_dispatch());
+	_pipeline = _dev->get_device().createGraphicsPipeline(nullptr, gpci, allocator, _dev->get_dispatch()).value;
 }
 
 void vge::GraphicsPipeline::AddVertexBinding(uint32_t stride, vk::VertexInputRate rate)
@@ -172,5 +172,5 @@ void vge::ComputePipeline::Create()
 		throw std::runtime_error("Compute shader need 1 stage");
 	}
 	cpci.stage = _shaders[0];
-	_pipeline = _dev->get_device().createComputePipeline(nullptr, cpci, allocator, _dev->get_dispatch());
+	_pipeline = _dev->get_device().createComputePipeline(nullptr, cpci, allocator, _dev->get_dispatch()).value;
 }

